### PR TITLE
Add support of AWS session token

### PR DESCRIPTION
**Pull request by ArinGhazarian**
Closes: #843 
Fixes: #744 

- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked
- [x] Docs updated (or issue created)
- [x] New package licenses are added to `ThirdPartyNotices.txt` (if applicable)

### Description

This PR adds support of the long awaited AWS session token authentication to `bbs2gh migrate-repo` and `gei migrate-repo` commands. Additionally the AWS region can also be passed into both `migrate-repo` and `generate-script` for `bbs2gh` and `gei`. 

This PR naturally obsoletes #826, #731 and #824. Thanks @leonli-xero and @derwasp for your invaluable contributions.

### CODE REVIEW

File: RELEASENOTES.md
Comment (timrogers) [inline RELEASENOTES.md:2]:
```suggestion
- Add support for authenticating with AWS session tokens when using AWS S3 for archive upload in `gh gei` and `gh bbs2gh`. When specifying a session, the AWS region must also be specified.
```

File: src/OctoshiftCLI.Tests/bbs2gh/Handlers/GenerateScriptCommandHandlerTests.cs
Comment (timrogers) [inline src/OctoshiftCLI.Tests/bbs2gh/Handlers/GenerateScriptCommandHandlerTests.cs:348]: Should we have a test which covers what happens if you don't submit a region? (I think we said that the region would be required for `generate-script`, but not `migrate-repo`.)

File: src/OctoshiftCLI.Tests/gei/Handlers/GenerateScriptCommandHandlerTests.cs
Comment (timrogers) [inline src/OctoshiftCLI.Tests/gei/Handlers/GenerateScriptCommandHandlerTests.cs:1412]: Should we have a test which covers what happens if you don't submit a region? (I think we said that the region would be required for `generate-script`, but not `migrate-repo`.)

File: src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs
Comment (timrogers) [inline src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs:519]: > Specifying an AWS region with the --aws-region argument or AWS_REGION environment variable is currently not required, but will be required in a future release. Defaulting to us-east-1.

File: src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs
Comment (timrogers) [inline src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs:524]:
```suggestion
            throw new OctoshiftCliException("The AWS S3 bucket name must be provided with --aws-bucket-name if other AWS S3 upload options are set.");
```

File: src/gei/Handlers/MigrateRepoCommandHandler.cs
Comment (timrogers) [inline src/gei/Handlers/MigrateRepoCommandHandler.cs:489]: > Specifying an AWS region with the --aws-region argument or AWS_REGION environment variable is currently not required, but will be required in a future release. Defaulting to us-east-1.

File: src/gei/Commands/MigrateRepoCommand.cs
Comment (timrogers) [inline src/gei/Commands/MigrateRepoCommand.cs:111]:
```suggestion
            Description = "If using AWS, the AWS region. If not provided, it will be read from AWS_REGION environment variable. Defaults to us-east-1 if neither the argument nor the environment variable is set. In a future release, you will be required to set an AWS region if using AWS S3 as your blob storage provider."
```

File: src/gei/Commands/GenerateScriptCommand.cs
Comment (timrogers) [inline src/gei/Commands/GenerateScriptCommand.cs:107]:
```suggestion
            Description = "If using AWS, the AWS region. If not provided, it will be read from AWS_REGION environment variable. Defaults to us-east-1 if neither the argument nor the environment variable is set. In a future release, you will be required to set an AWS region if using AWS S3 as your blob storage provider."
```

File: src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs
Comment (timrogers) [inline src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs:517]: Can we add a validation that requires the user to set their AWS region if they provide a session token? We're only trying to preserve backwards compatibility for old scripts really, and if you specify a session token, then it's a good time to force the region too I think.

File: src/OctoshiftCLI.Tests/bbs2gh/Handlers/MigrateRepoCommandHandlerTests.cs
Comment (timrogers) [inline src/OctoshiftCLI.Tests/bbs2gh/Handlers/MigrateRepoCommandHandlerTests.cs:900]: Should we also add a test for the case where a session token is provided but a region is not provided?

**Comment by timrogers** [comment]
@ArinGhazarian Thanks for smashing this out ❤️

**Review by dylan-smith — CHANGES_REQUESTED** [review]
I gave it a look over and I don't have anything to add beyond what @timrogers suggested. I agree that we should make region required on migrate-repo if session-token is provided. And that region should be required on generate-script (if using AWS). And that we should have tests for both with and without region for these scenarios.

File: src/OctoshiftCLI.Tests/bbs2gh/Handlers/GenerateScriptCommandHandlerTests.cs
Comment (ArinGhazarian) [reply src/OctoshiftCLI.Tests/bbs2gh/Handlers/GenerateScriptCommandHandlerTests.cs:348]: In `generate-script` so far we have done very minimal validations the actual validations are going to happen in `migrate-repo` so there is no need to repeat them in `generate-script` too unless it's crucial to `generate-script` itself.

File: src/OctoshiftCLI.Tests/gei/Handlers/GenerateScriptCommandHandlerTests.cs
Comment (ArinGhazarian) [reply src/OctoshiftCLI.Tests/gei/Handlers/GenerateScriptCommandHandlerTests.cs:1412]: see [this](https://github.com/github/gh-gei/pull/844#discussion_r1123632495)

File: src/OctoshiftCLI.Tests/bbs2gh/Handlers/MigrateRepoCommandHandlerTests.cs
Comment (ArinGhazarian) [reply src/OctoshiftCLI.Tests/bbs2gh/Handlers/MigrateRepoCommandHandlerTests.cs:900]: Not now, because the region is actually optional in our code so it's totally valid to provide the session token but not the region. that being said I can add the test and make sure that we're printing the log message and it future once we made the region required we can just update the test to check for the validation.

File: src/OctoshiftCLI.Tests/bbs2gh/Handlers/GenerateScriptCommandHandlerTests.cs
Comment (ArinGhazarian) [reply src/OctoshiftCLI.Tests/bbs2gh/Handlers/GenerateScriptCommandHandlerTests.cs:348]: Oh I see, it makes perfect sense to make region required if session token is passed! I will make that change in `migrate-repo` and add the validation. I still don't think we need to add extra validations in `generate-script` though.

File: RELEASENOTES.md
Comment (ArinGhazarian) [reply RELEASENOTES.md]: adressed in 98de164b93e6c5ad0a9f24c52673cd80b4f9a7fc

File: src/OctoshiftCLI.Tests/bbs2gh/Handlers/GenerateScriptCommandHandlerTests.cs
Comment (ArinGhazarian) [reply src/OctoshiftCLI.Tests/bbs2gh/Handlers/GenerateScriptCommandHandlerTests.cs:348]: Addressed in 98de164b93e6c5ad0a9f24c52673cd80b4f9a7fc

File: src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs
Comment (ArinGhazarian) [reply src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs]: Addressed in 98de164b93e6c5ad0a9f24c52673cd80b4f9a7fc

File: src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs
Comment (ArinGhazarian) [reply src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs]: Addressed in 98de164b93e6c5ad0a9f24c52673cd80b4f9a7fc

File: src/gei/Handlers/MigrateRepoCommandHandler.cs
Comment (ArinGhazarian) [reply src/gei/Handlers/MigrateRepoCommandHandler.cs]: Addressed in 98de164b93e6c5ad0a9f24c52673cd80b4f9a7fc

File: src/gei/Commands/MigrateRepoCommand.cs
Comment (ArinGhazarian) [reply src/gei/Commands/MigrateRepoCommand.cs]: Addressed in 98de164b93e6c5ad0a9f24c52673cd80b4f9a7fc

File: src/gei/Commands/GenerateScriptCommand.cs
Comment (ArinGhazarian) [reply src/gei/Commands/GenerateScriptCommand.cs]: Addressed in 98de164b93e6c5ad0a9f24c52673cd80b4f9a7fc

File: src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs
Comment (ArinGhazarian) [reply src/bbs2gh/Handlers/MigrateRepoCommandHandler.cs:523]: Addressed in 98de164b93e6c5ad0a9f24c52673cd80b4f9a7fc

**Comment by ArinGhazarian** [comment]
@dylan-smith and @timrogers, I addressed all your comments with one exception. For `generate-script` command we haven't added exhaustive validations as we did for `migrate-repo` command unless required for the `generate-script`'s functionality itself. 
I don't think we need to duplicate the `migrate-repo` validations for `generate-script` but If you guys think that we should, then it'd be better to address it in a separate issue because the validations will be almost identical to `migrate-repo` and we probably need to extract them into its own class to be DRY.

File: src/Octoshift/AwsApi.cs
Comment (timrogers) [inline src/Octoshift/AwsApi.cs:33]: Will this throw an exception if an invalid region name is specified? I just want to make sure I understand our failure modes!

File: src/Octoshift/AwsApi.cs
Comment (timrogers) [inline src/Octoshift/AwsApi.cs:34]: ❤️

**Review by timrogers — APPROVED** [review]
Looking good to me ❇️ But let's wait for a review from Dylan too.

**Comment by timrogers** [comment]
> @dylan-smith and @timrogers, I addressed all your comments with one exception. For `generate-script` command we haven't added exhaustive validations as we did for `migrate-repo` command unless required for the `generate-script`'s functionality itself. I don't think we need to duplicate the `migrate-repo` validations for `generate-script` but If you guys think that we should, then it'd be better to address it in a separate issue because the validations will be almost identical to `migrate-repo` and we probably need to extract them into its own class to be DRY.

My personal view is that we should have more exhaustive validations at the script generation stage - but maybe I'm being too conservative, so I'd love to hear Dylan's thoughts. In any event, it can be a follow-up PR!

**Review by dylan-smith — APPROVED** [review]

**Comment by dylan-smith** [comment]
> > @dylan-smith and @timrogers, I addressed all your comments with one exception. For `generate-script` command we haven't added exhaustive validations as we did for `migrate-repo` command unless required for the `generate-script`'s functionality itself. I don't think we need to duplicate the `migrate-repo` validations for `generate-script` but If you guys think that we should, then it'd be better to address it in a separate issue because the validations will be almost identical to `migrate-repo` and we probably need to extract them into its own class to be DRY.
> 
> My personal view is that we should have more exhaustive validations at the script generation stage - but maybe I'm being too conservative, so I'd love to hear Dylan's thoughts. In any event, it can be a follow-up PR!

I agree, but I'm not sure what we're missing. `generate-script` only takes `--aws-bucket-name` and `--aws-region` and region is optional so there's not a whole lot of validation needed. What am I missing?

**Comment by timrogers** [comment]
> > > @dylan-smith and @timrogers, I addressed all your comments with one exception. For `generate-script` command we haven't added exhaustive validations as we did for `migrate-repo` command unless required for the `generate-script`'s functionality itself. I don't think we need to duplicate the `migrate-repo` validations for `generate-script` but If you guys think that we should, then it'd be better to address it in a separate issue because the validations will be almost identical to `migrate-repo` and we probably need to extract them into its own class to be DRY.
> > 
> > 
> > My personal view is that we should have more exhaustive validations at the script generation stage - but maybe I'm being too conservative, so I'd love to hear Dylan's thoughts. In any event, it can be a follow-up PR!
> 
> I agree, but I'm not sure what we're missing. `generate-script` only takes `--aws-bucket-name` and `--aws-region` and region is optional so there's not a whole lot of validation needed. What am I missing?

That's a good point. I think that probably is everything, actually. I'll do some more testing to confirm, and we can always have a follow-up PR.

**Comment by ArinGhazarian** [comment]
> > > @dylan-smith and @timrogers, I addressed all your comments with one exception. For `generate-script` command we haven't added exhaustive validations as we did for `migrate-repo` command unless required for the `generate-script`'s functionality itself. I don't think we need to duplicate the `migrate-repo` validations for `generate-script` but If you guys think that we should, then it'd be better to address it in a separate issue because the validations will be almost identical to `migrate-repo` and we probably need to extract them into its own class to be DRY.
> > 
> > 
> > My personal view is that we should have more exhaustive validations at the script generation stage - but maybe I'm being too conservative, so I'd love to hear Dylan's thoughts. In any event, it can be a follow-up PR!
> 
> I agree, but I'm not sure what we're missing. `generate-script` only takes `--aws-bucket-name` and `--aws-region` and region is optional so there's not a whole lot of validation needed. What am I missing?

Well for AWS it only takes those two but for validating them we need to do extra stuff, for example we don't pass secrets via args in `generate-script` so we need to read the env. variables to do the validations. Also there are other args that we haven't validated and if we decide to go this route we need to add validations for them too that eventually leads to basically duplicating the validation for `migrate-repo`. Also there's always a possibility that they modify the script and add/remove stuff.

**Comment by dylan-smith** [comment]
> 
> Well for AWS it only takes those two but for validating them we need to do extra stuff, for example we don't pass secrets via args in `generate-script` so we need to read the env. variables to do the validations. Also there are other args that we haven't validated and if we decide to go this route we need to add validations for them too that eventually leads to basically duplicating the validation for `migrate-repo`. Also there's always a possibility that they modify the script and add/remove stuff.

I would want generate-script to work without requiring those env vars to be set. They have to be set before the script can be run, but it should be able to be generated without credentials set.

What other args haven't we validated?

**Comment by ArinGhazarian** [comment]
> I would want generate-script to work without requiring those env vars to be set. They have to be set before the script can be run, but it should be able to be generated without credentials set.
> 

That's exactly my point (and why I think it'd be overkill to do validations for `generate-script`). Most of those validations need the env. variable to be set. For example `aws-region` is only required (for now) if the `AWS_SESSION_TOKEN` is set. And once we made the aws region required we're gonna mark it so in the command so we'll get the validation for free.

> What other args haven't we validated?
For example the ssh and smb options in `bbs2gh`, etc.

**Comment by dylan-smith** [comment]
In generate-script I thought we made aws-region required if aws-bucket was set, but I see now we didn't. I'm struggling to see what other validations would be necessary in gei generate-script. I wouldn't want the session token to be needed to be provided at script generation time.

For bbs2gh generate-script I just noticed we have ZERO validation. That's a problem we should fix.

File: src/Octoshift/AwsApi.cs
Comment (ArinGhazarian) [reply src/Octoshift/AwsApi.cs:33]: @timrogers I am not sure how I missed this comment, but this is a very valid question. I was naturally under the impression that `GetBySystemName()` would throw if the region string was invalid but interestingly that's not the case! `GetBySystemName()` always succeeds even for an invalid region name! Not sure why they don't validate it! The migration fails eventually but we should def. validate and throw if the region is invalid. Fail fast, fail early 😄 
I am going to put together a quick fix for this. Thanks for pointing this out 🙏

Comment (ArinGhazarian) [reply src/Octoshift/AwsApi.cs:33]: PR for the fix --> https://github.com/github/gh-gei/pull/872

**Pull request merged**
